src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -3494,17 +3496,12 @@ public String getKeyspaceReplicationInfo(String keyspaceName)
     public Set<InetAddressAndPort> getLiveRingMembers(boolean excludeDeadStates)
     {
         Set<InetAddressAndPort> ret = new HashSet<>();
-        for (InetAddressAndPort ep : Gossiper.instance.getLiveMembers())
+        for (InetAddressAndPort ep : getTokenMetadata().getAllRingMembers())
         {
-            if (excludeDeadStates)
-            {
-                EndpointState epState = Gossiper.instance.getEndpointStateForEndpoint(ep);
-                if (epState == null || Gossiper.instance.isDeadState(epState))
-                    continue;
-            }
+            if (excludeDeadStates && !IFailureDetector.instance.isAlive(ep))
+                continue;
 
-            if (tokenMetadata.isMember(ep))
-                ret.add(ep);
+            ret.add(ep);
```

Comment:
It is hard for me to validate that the change is equivalent. I understand the idea.